fish-rust/src/builtins/functions.rs

```diff
@@ -0,0 +1,413 @@
+                "%ls: Expected generic | variable | signal | exit | job-id for --handlers-type\n",
+                cmd
+            ));
+            return STATUS_INVALID_ARGS;
+        }
+        event::print(streams, &opts.handlers_type);
+        return STATUS_CMD_OK;
+    }
+
+    if opts.query && args.is_empty() {
+        return STATUS_CMD_ERROR;
+    }
+
+    if opts.list || args.is_empty() {
+        let mut names = function::get_names(opts.show_hidden);
+        names.sort();
+        if !streams.out_is_redirected && unsafe { isatty(STDOUT_FILENO) == 1 } {
+            let mut buff = WString::new();
+            let mut first: bool = true;
+            for name in names {
+                if !first {
+                    buff.push_utfstr(L!(", "));
+                }
+                buff.push_utfstr(&name);
+                first = false;
+            }
+            // TODO: reformat_for_screen
+            streams.out.append(buff);
+        } else {
+            for name in names {
+                streams.out.append(name + "\n");
+            }
+        }
+        return STATUS_CMD_OK;
+    }
+
+    if opts.copy {
+        if args.len() != 2 {
+            streams.err.append(wgettext_fmt!(
+                "%ls: Expected exactly two names (current function name, and new function name)\n",
+                cmd
+            ));
+            builtin_print_error_trailer(parser, streams, cmd);
+            return STATUS_INVALID_ARGS;
+        }
+        let current_func = args[0];
+        let new_func = args[1];
+
+        if !function::exists(current_func, parser) {
+            streams.err.append(wgettext_fmt!(
+                "%ls: Function '%ls' does not exist\n",
+                cmd,
+                current_func
+            ));
+            builtin_print_error_trailer(parser, streams, cmd);
+            return STATUS_CMD_ERROR;
+        }
+
+        if !valid_func_name(new_func) || parser_keywords_is_reserved(new_func) {
+            streams.err.append(wgettext_fmt!(
+                "%ls: Illegal function name '%ls'\n",
+                cmd,
+                new_func
+            ));
+            builtin_print_error_trailer(parser, streams, cmd);
+            return STATUS_INVALID_ARGS;
+        }
+
+        if function::exists(new_func, parser) {
+            streams.err.append(wgettext_fmt!(
+                "%ls: Function '%ls' already exists. Cannot create copy '%ls'\n",
+                cmd,
+                new_func,
+                current_func
+            ));
+            builtin_print_error_trailer(parser, streams, cmd);
+            return STATUS_CMD_ERROR;
+        }
+        if function::copy(current_func, new_func.into(), parser) {
+            return STATUS_CMD_OK;
+        }
+        return STATUS_CMD_ERROR;
+    }
+
+    let mut res: c_int = STATUS_CMD_OK?;
+
+    let mut first = true;
+    for arg in args.iter() {
+        if opts.query {
+            if !function::exists(arg, parser) {
```

Comment:
This is a slight change in that we now check function::exists for query, instead of seeing that the properties don't exist.